components/cop_datatype/src/def/eval_type.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+/// Function implementations' parameter data types.
+///
+/// It is similar to the `EvalType` in TiDB, but doesn't provide type `Timestamp`, which is
+/// handled by the same type as `DateTime` here instead of a new type. Also, `String` is
+/// called `Bytes` here to be less confusing.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum EvalType {
+    Int,
+    Real,
+    Decimal,
+    Bytes,
+    DateTime,
+    Duration,
+    Json,
+}
+
+impl fmt::Display for EvalType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+impl ::std::convert::TryFrom<::FieldTypeTp> for EvalType {
+    type Error = ::Error;
+
+    fn try_from(tp: ::FieldTypeTp) -> Result<Self, ::Error> {
+        let eval_type = match tp {
+            ::FieldTypeTp::Tiny
+            | ::FieldTypeTp::Short
+            | ::FieldTypeTp::Int24
+            | ::FieldTypeTp::Long
+            | ::FieldTypeTp::LongLong
+            | ::FieldTypeTp::Year => EvalType::Int,
+            ::FieldTypeTp::Float | ::FieldTypeTp::Double => EvalType::Real,
+            ::FieldTypeTp::NewDecimal => EvalType::Decimal,
+            ::FieldTypeTp::Timestamp | ::FieldTypeTp::Date | ::FieldTypeTp::DateTime => {
```

Comment:
Why TiDB use a `ETimestamp` type for `FieldTypeTp::Timestamp`?

Reply:
https://github.com/pingcap/parser/blob/master/types/field_type.go#L87

Reply:
It means E(val)T(Type)Timestamp. There is already a constant Timestamp (FieldType) in the same package and golang doesn't allow name duplications. In TiKV, we don't need the ET prefix.

Reply:
No, I mean why TiDB use two different types for `Date`, `DateTime` and `Timestamp`?

```
	case mysql.TypeDate, mysql.TypeDatetime:
		return ETDatetime
	case mysql.TypeTimestamp:
		return ETTimestamp
```